### ISSUE TITLE
Optimizing engine.dispatchEvent

### ISSUE THREAD

**Issue by NDStrahilevitz** [aquasecurity/tracee#1372]
During the ongoing performance research of tracee I have found the following `convT2E` code path in engine.dispatchEvent:

![image](https://user-images.githubusercontent.com/22661609/150386593-feac49dc-4c27-465b-846d-235aca615af8.png)

This code path results from this code block:
```
case strings.Contains(reflect.TypeOf(s).String(), "rego"):
		if engine.parsedEvents {
			pe, err := ToParsedEvent(event)
			if err != nil {
				engine.logger.Printf("error converting tracee event to OPA ast.Value: %v", err)
				return
			}
			engine.signatures[s] <- pe
		} else {
			engine.signatures[s] <- event
		}
```

the reflection method here converts the type to interface and then back, which as you can see in the flame graph takes a signficant chunk of the entire function call.

Me and @mtcherni95 have thought of 2 strategies to change this:

1. Add an enum field `SignatureType` to the `SignatureMetadata` struct which will look as such:
```
type SignatureType int

const (
	UndefinedSigType SignatureType = iota
	RegoSigType
	GoSigType
)
```

We could then add a case in the switch case above to check against RegoSigType, and later we can remove the unperformant code path once we decide enough time has passed to adopt this struct API.
2. Add a method to the `Signature` interface like so:
```
...
ParseEvent(event external.Event) external.Event
```

and then change the `ParsedEvent` type to embed `external.Event` instead of containing it. Thus we could remove the switch altogether and use the signature's implementation of event parsing to send to the signature channel in the end.

I Would love to hear thoughts about the existing approaches or if there are other approaches that haven't been considered :)

**Comment by danielpacak** [comment]
## TL;DR;

IMO, since we use OPA > v0.30.0 we may consider removing all that ParsedEvent and ToParsedEvent logic at all. Today we know that it wasn't the main performance issue with OPA that we had.

Beyond that, we do have leaking abstractions in the Engine. Ideally, the Engine should not be aware of OPA or Go signature implementation details. It should act on the Event and Signature abstraction. We broke this abstraction by introducing ParsedEvent and passing OPA specific flags with hope to make it more performant, but it never happened.

## Why we added ParsedEvent and the --rego-enable-parsed-events flag?

Some time back when we did a similar profiling for tracee-rules we found out that OPA Go SDK during eval is using the `ast.InterfaceToValue(u)` method that is "expensive". What's more it was invoked by each Rego signature. (We were considering tens or more Rego signatures.) That's why we introduced the `--rego-enable-parsed-events` flag to experiment with parsed events and do such expensive event processing once at dispatch time no matter how many Rego signatures you have. Another requirement was to keep the Signature interface (and implementations) backward compatible and that's why you see the ParsedEvent struct, which is only relevant for OPA.

We shared our findings with OPA maintainers and they have implemented a few improvements around that functionality to the point where the ToParsedEvent methods is no longer required. (See the TODO comment in the ToParseEvent method.)

**Comment by NDStrahilevitz** [comment]
So as you see it @danielpacak , a PR just removing that logic would suffice? If so then I think there's no need for the solutions initially proposed. 

I will suggest another simple fix anyways: we could also just move that whole switch case inside a `if engine.parsedEvent` instead of having that if statement inside the switch case, since that's the whole reason for that logic anyways.

**Comment by NDStrahilevitz** [comment]
So after further investigation the reason for `runtime.convT2E` calls are not due to the switch case reflection (although they do contribute), but the fact that the engine signatures are defined as channels that receive `types.Event` which is just `interface{}`. That means that each event dispatched triggers this runtime function. To solve this we need to eliminate use of the generic `types.Event` entirely in tracee-rules.

@danielpacak Am I correct in thinking that `types.Event` is `interface{}` due to the parsed event logic is or is there a larger design choice behind it? If so as part of removing parsed events we can also start replacing `types.Event` with tracee's Event type. This is a change that goes somewhat deeper down so if we go down that route I think we should open a new issue tracking both these changes.

**Comment by danielpacak** [comment]
ParsedEvent has nothing to do with generic event of type `interface{}`. ParsedEvent just embedded Event and added the ast.Value to save computed state before it's passed to Rego signatures.

IIRC we wanted to support other event types and that was the main reason behind very generic `interface{}`. Maybe @itaysk can provide more context here, but I agree if we can avoid casting in so many places we should consider refactor.

**Comment by yanivagman** [comment]
> @danielpacak Am I correct in thinking that `types.Event` is `interface{}` due to the parsed event logic is or is there a larger design choice behind it? If so as part of removing parsed events we can also start replacing `types.Event` with tracee's Event type. This is a change that goes somewhat deeper down so if we go down that route I think we should open a new issue tracking both these changes.

The reason we chose to implement the event as an interface is because we wanted tracee-rules to be able to handle not only tracee-ebpf events, but possibly other events in the future (e.g. kubernetes logs).

**Comment by NDStrahilevitz** [comment]
I'd like to continue the discussion of this in #1431, this issue itself missed the larger point so i'm closing it.
